### PR TITLE
KAFKA-17540: Create script for updating a reference of latest cached trunk commit

### PR DESCRIPTION
**Pull request by fonsdant**

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -83,6 +83,14 @@ jobs:
         run: python .github/scripts/checkstyle.py
         env:
           GITHUB_WORKSPACE: ${{ github.workspace }}
+      - name: Update Floating 'cached' Tag on Successful Trunk Build
+        if: success()
```
Comment (mumrah) [inline .github/workflows/build.yml]: This workflow is used by PRs as well as trunk. We need to make sure we're only tagging the commits on trunk. You could do that here with an additional condition, or add a trunk-only step to ci.yml

**Comment by fonsdant** [comment]
Hi, @mumrah! I have made a test to check if it works as expected. You can check it: https://github.com/fonsdant/github-actions.

To test, I have created two jobs: one for a successful build (and then tag update) and another for a failed build (no tag update). I have needed to give read and write permission to GitHub Action to perform the tag creation.

- [Failed build because of read and write permission](https://github.com/fonsdant/github-actions/actions/runs/10902518962)
- [Completed build for creating cached tag](https://github.com/fonsdant/github-actions/actions/runs/10902555717)
- [Completed build for PR](https://github.com/fonsdant/github-actions/actions/runs/10902593757)
- [Completed build for merged PR (update tag)](https://github.com/fonsdant/github-actions/actions/runs/10902740842)

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -83,6 +83,14 @@ jobs:
         run: python .github/scripts/checkstyle.py
         env:
           GITHUB_WORKSPACE: ${{ github.workspace }}
+      - name: Update Floating 'cached' Tag on Successful Trunk Build
+        if: github.ref == 'refs/heads/trunk' && success()
+        run: |
+          git fetch --tags
```
Comment (mumrah) [inline .github/workflows/build.yml]: I think you can fetch a single tag here to save a bit of time.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -83,6 +83,14 @@ jobs:
...
+          git tag -f cached $GITHUB_SHA
```
Comment (mumrah) [inline .github/workflows/build.yml]: When we create the new tag, can we add a message like "Automatically generated after running workflow 1234. https://github.com/fonsdant/github-actions/actions/runs/10902593757" or similar? This will help us if we need to debug caching issues

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -83,6 +83,14 @@ jobs:
         run: python .github/scripts/checkstyle.py
         env:
           GITHUB_WORKSPACE: ${{ github.workspace }}
+      - name: Update Floating 'cached' Tag on Successful Trunk Build
```
Comment (mumrah) [inline .github/workflows/build.yml]: I like the thorough description, but I think the "on Successful Trunk Build" qualifier is not needed. In the workflow run UI, we'll see this step skipped in all cases except on successful trunk builds :)

**Review by mumrah — COMMENTED** [review]
Thanks @fonsdant! Left a few comments.

I also wonder if we should put this in the CI Complete workflow. Though unlikely, it is possible that the build succeeds, but writing to the cache fails in the Post Gradle Setup step. If this happened, we would have updated the tag prematurely. 

If we add a "update-cached-tag" job in CI Complete, we can trigger it to only run when a CI workflow on trunk is successful (which would mean the cache was updated).

File: .github/workflows/build.yml (hunk 2 above)
Comment (fonsdant) [reply .github/workflows/build.yml]: Sure!

File: .github/workflows/build.yml (hunk 3 above)
Comment (fonsdant) [reply .github/workflows/build.yml]: Updated!

File: .github/workflows/build.yml (hunk 4 above)
Comment (fonsdant) [reply .github/workflows/build.yml]: Agree, have updated it!

**Comment by fonsdant** [comment]
> Thanks @fonsdant! Left a few comments.
> 
> I also wonder if we should put this in the CI Complete workflow. Though unlikely, it is possible that the build succeeds, but writing to the cache fails in the Post Gradle Setup step. If this happened, we would have updated the tag prematurely.
> 
> If we add a "update-cached-tag" job in CI Complete, we can trigger it to only run when a CI workflow on trunk is successful (which would mean the cache was updated).

Moved!

File: .github/workflows/ci-complete.yml [hunk 5]
```diff
@@ -118,3 +118,17 @@ jobs:
           description: 'The build scan was successfully published'
           context: 'Gradle Build Scan / Java ${{ matrix.java }}'
           state: 'success'
+
+  update-cached-tag:
+    # Skip this workflow if the CI run was skipped or cancelled
+    if: (github.event.workflow_run.conclusion == 'success' || github.event.workflow_run.conclusion == 'failure')
```
Comment (mumrah) [inline .github/workflows/ci-complete.yml]: In this case, we only want to run if the workflow was successful.

**Review by mumrah — COMMENTED** [review]
One small thing remaining. 

Also, can you see if signed tags are possible here? That might be a nice thing to include.

File: .github/workflows/ci-complete.yml (hunk 5 above)
Comment (fonsdant) [reply .github/workflows/ci-complete.yml]: Oh, thanks! I will fix it!

**Comment by fonsdant** [comment]
> One small thing remaining.
> 
> Also, can you see if signed tags are possible here? That might be a nice thing to include.

Yes, they are possible but require a GPG key. Is the purpose of the sign to be the same as the commits? If so, maybe we could add a `Signed-off-by` trailer at the end of the description. But I think it will require an automation, it is not out-of-the-box.

**Comment by mumrah** [comment]
> Is the purpose of the sign to be the same as the commits? 

No, the purpose is just to have a signature on the tag itself. When Github merges PRs, it will sign the resulting commit with its own gpg key. I was hoping that doing `git tag -s` in the action would do the same for tags. It should be possible to test it on your fork.

**Comment by mumrah** [comment]
I think tag signing with `git` may not work as expected here. It may be possible to use the API to create a signed tag. https://docs.github.com/en/rest/git/tags?apiVersion=2022-11-28

**Comment by fonsdant** [comment]
> I think tag signing with `git` may not work as expected here.

I have made a test. Only to confirm: is it not work as expected because of [this](https://github.com/fonsdant/github-actions/actions/runs/10940325749/job/30372492274#step:4:25)? Git could not determine GitHub bot's email.

Maybe we could use `gh` CLI to create the tag. [GitHub CLI is preinstalled on all GitHub-hosted runners](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/using-github-cli-in-workflows)

**Comment by mumrah** [comment]
Yes, using `gh` should work. The docs link I posted has an example of creating a tag with `gh api`

```
gh api \
  --method POST \
  -H "Accept: application/vnd.github+json" \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  /repos/OWNER/REPO/git/tags \
   -f "tag=v0.0.1" -f "message=initial version" -f "object=c3d0be41ecbe669545ee3e94d31ed9a4bc91ee3c" -f "type=commit" -f "tagger[name]=Monalisa Octocat" -f "tagger[email]=octocat@github.com" -f "tagger[date]=2011-06-17T14:53:35-07:00"
```

We're using this tool to create custom statuses on PRs here https://github.com/apache/kafka/blob/trunk/.github/actions/gh-api-update-status/action.yml#L55-L59

**Comment by fonsdant** [comment]
Oh! I have noted it fails when tag ref already exists. I am working on fix it!

**Comment by mumrah** [comment]
@fonsdant first off, thanks for all the work on this. I was pondering this yesterday and had an idea. Instead of managing a tag, I think we can use a separate branch, something like `trunk-cached`. This branch just contains the commits from `trunk` that are included in the latest GitHub cache. 

In this case, the CI becomes simply

```
git checkout trunk-cached
git merge SHA
git push origin trunk-cached
```

This will improve the developer experience somewhat since it doesn't require explicitly fetching tags. Just a simple `git fetch origin` will pull it down.

Can you give this a try?

**Comment by fonsdant** [comment]
I am happy to help! :)

Sure! I will give it a try. It seems simpler for me too!

**Comment by fonsdant** [comment]
Hi, @mumrah! I have made some modifications. I have used `switch` as it seems more semantic. Instead of merge, I have chosen to use `reset --hard` to prevent any conflicts that `merge` could raise. How about it?

File: .github/workflows/ci-complete.yml [hunk 6]
```diff
@@ -118,3 +118,16 @@ jobs:
...
           state: 'success'
+
+  update-cached-tag:
+    # Only run this job if the previous workflow run was successful
+    if: github.event.workflow_run.conclusion == 'success'
+    runs-on: ubuntu-latest
+    steps:
+      - name: Update trunk-cached branch with trunk
+        run: |
+          git switch trunk-cached || git switch -c trunk-cached
```
Comment (mumrah) [inline .github/workflows/ci-complete.yml]: Do we need to fetch trunk-cached first? Or does the checkout step fetch everything?

Comment (fonsdant) [reply .github/workflows/ci-complete.yml]: As we are doing a `reset` here, it is not needed to do `fetch`. Even if we do `fetch`, the fetched trunk-cached branch will be reset to `github.sha` as well as the unfetched trunk-cached branch. So the fetched trunk-cached does not change the result.

In a way, the `github.sha` is already our "fetched branch" or "reference", which the trunk-cached branch should point to. And `push -f` guarantees the update will be performed without fast-forward or similar issues.

I have done two tests for this. You can check them [here](https://github.com/fonsdant/github-actions/actions/runs/10968527602/job/30459981606#step:3:11) and [here](https://github.com/fonsdant/github-actions/actions/runs/10968538249/job/30460008394#step:3:11).

This is the `git log` result:

```
$ git log --oneline 
e8da285 (HEAD -> trunk, origin/trunk-cached, origin/trunk) update ble.txt
d9c105e minimal impl
425d23e replace sha with origin trunk branch
...
```

File: .github/workflows/ci-complete.yml [hunk 7]
```diff
@@ -118,3 +118,16 @@ jobs:
           description: 'The build scan was successfully published'
           context: 'Gradle Build Scan / Java ${{ matrix.java }}'
           state: 'success'
+
+  update-cached-tag:
```
Comment (mumrah) [inline .github/workflows/ci-complete.yml]: this should probably be `update-trunk-cached` now

File: .github/workflows/ci-complete.yml (hunk 6 above)
Comment (mumrah) [reply .github/workflows/ci-complete.yml]: Thanks for the explanation. Looks like we're always doing `git switch -c` to create a local branch, so maybe we don't need the first `git switch` ?

**Review by mumrah — COMMENTED** [review]
Thinking more about it, I'm a little hesitant about the force push. There's nothing semantically wrong with it, but it just gives bad vibes 😅. Can we do: fetch, merge --ff-only, push instead? 

If somehow `trunk` has diverged from `trunk-cached`, we would have to manually fixup `trunk-cached` which doesn't seem like a big deal. This would only happen if trunk itself had its history rewritten which is not something that has ever happened AFAIK.

File: .github/workflows/ci-complete.yml (hunk 6 above)
Comment (fonsdant) [reply .github/workflows/ci-complete.yml]: You are right! I will remove the `git switch`. Thanks!!

**Comment by fonsdant** [comment]
> Can we do: fetch, merge --ff-only, push instead?

It seems this approach is not so out-of-the-box. In my tests, Git fails with `fatal: refusing to merge unrelated histories`: [here](https://github.com/apache/kafka/pull/17204#pullrequestreview-2319726159).

```
git switch -c trunk-cached
if git fetch origin trunk-cached; then
  git reset --hard origin/trunk-cached
else
  echo 'Remote ref does not exist'
fi
git merge ${{ github.sha }} # fail here
git push origin trunk-cached
```

Even if we add `--allow-unrelated-histories`, then the failure becomes a conflict merge: [here](https://github.com/fonsdant/github-actions/actions/runs/10977036891/job/30478533261#step:3:20). Curiously, I have no issues performing merge locally. I am trying to understand if I have made something wrong in remote repo.

**Comment by fonsdant** [comment]
How about using only a ref to point to the cached commit instead?

A tag isn't the best option, as described in the GitHub REST API documentation: _"A Git tag is similar to a Git reference, but the Git commit it points to never changes."_

A branch doesn’t seem ideal either, since we're not using it for its intended features. We’ve created the branch solely to mark the cached commit, and there’s also the risk of merge conflicts.

So, a ref seems to be a great option. It acts as an "alias" for the cached commit and can be updated to point to a new cached commit.

What do you think?

**Comment by mumrah** [comment]
@fonsdant, just updating the ref on the remote (GitHub) would indeed work.

In Git, a ref is simply a text file with a SHA in it. Conventionally, tags are immutable since they are typically used to refer to a point in the history (like a release), but as you know we can update them with the --force flag.

A branch is actually just a ref as well. In the .git/refs directory, the branch refs are in the "heads" directory. On a fresh clone of Kafka, we see:

```
.git/refs
.git/refs/heads
.git/refs/heads/trunk
.git/refs/tags
.git/refs/remotes
.git/refs/remotes/origin
.git/refs/remotes/origin/HEAD
```

(note that other branches/tags are compacted into a "packed_refs" file)

The file `.git/refs/heads/trunk` contains the SHA of trunk's HEAD commit at the time I fetched it. When I later do "git fetch", the commit objects will be updated (they are stored elsewhere) and the `.git/refs/heads/trunk` file will be overwritten with the new SHA for `trunk`.

> So, a ref seems to be a great option. It acts as an "alias" for the cached commit and can be updated to point to a new cached commit.

You've described a branch :) 

---

Looking at the "creating a ref" example from the API docs:

```
gh api \
  --method POST \
  -H "Accept: application/vnd.github+json" \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  /repos/OWNER/REPO/git/refs \
   -f "ref=refs/heads/featureA" -f "sha=aa218f56b14c9653891f9e74264a383fa43fefbd"
```

we can see they are actually creating "refs/heads/featureA" which is a branch. 

The approach we were taking using checkout/switch and merge was really just a roundabout way of updating the `.git/refs/heads/trunk-cached` file. 

If we use the API to update the "heads/trunk-cached" ref, I think we will accomplish our goal. I would like to _not_ do a forced update for now. We can re-address that if it becomes a problem in the future.

**Comment by fonsdant** [comment]
@mumrah, thanks for the explanation. I really appreciate it! :)

I will proceed modifying the script to use ref.

**Comment by fonsdant** [comment]
@mumrah, forgive me if I am being too repetitive or slow to develop. I am trying to achieve a great result to us and Kafka contributors.

I was afraid to proceed with merge approach as it have needed to use `--allow-unrelated-histories`  on my personal repo tests. But I have been afraid to proceed with ref approach too as I have noted that it is somehow "hided" from the git user, for example: when `git log` show tags and branches (local and remote), but not see the cached ref I have created.

So, I have decided to put a little more effort on branch approach, then I have found this PR: [fatal: refusing to merge unrelated histories](https://github.com/actions/checkout/issues/125). This explains why I am getting ["fatal: refusing to merge unrelated histories"](https://github.com/fonsdant/github-actions/actions/runs/11004390641/job/30555214700#step:3:28) and now ["fatal: Not possible to fast-forward, aborting"](https://github.com/fonsdant/github-actions/actions/runs/11004489567/job/30555506559#step:3:37).

According to [GitHub Checkout Action docs](https://github.com/marketplace/actions/checkout#Checkout-a-different-branch): _Only a single commit is fetched by default, for the ref/SHA that triggered the workflow._ As trunk is tringgering the workflow, only its head is fetched. I have achieved to fetch all trunk-based, but trunk no (even with `git fetch`). You can check the test on my personal repo: [here](https://github.com/fonsdant/github-actions/actions/runs/11004780392/job/30556346504#step:3:19). It's a problem for us to achieve our goal since...

```
old stuff -> A (cached) -> B -> C -> D -> HEAD
```

... we are going to lose B and C, and only merging D (latest commit fetched by checkout action).

I have made a test reversing the branches and setting trunk-cached as ref to checkout action. Therefore, the trunk history may be fetched and merged into trunk-cached (now, the one having only the latest commit fetched by checout action). This works as expected! See [here](https://github.com/fonsdant/github-actions/actions/runs/11005038170/job/30557041260) and [here](https://github.com/fonsdant/github-actions/actions/runs/11004838601/job/30556492241).

WDYT? :D

**Comment by mumrah** [comment]
In the [checkout action](https://github.com/actions/checkout) you can specify that you want a full checkout instead of sparse. 

```
    # Specifies whether to use cone-mode when doing a sparse checkout.
    # Default: true
    sparse-checkout-cone-mode: ''
```

This will let you checkout the workflow-triggering SHA and get its full history. Then, you can checkout ~trunk-clone~ `trunk-cached` and merge it like normal. Can you give that a try?

**Comment by fonsdant** [comment]
> In the [checkout action](https://github.com/actions/checkout) you can specify that you want a full checkout instead of sparse.

It seems to me we are already not doing a sparse checkout, since the default value for `sparse-checkout` is null and we are not passing any other value to it. See:

```
    # Do a sparse checkout on given patterns. Each pattern should be separated with
    # new lines.
    # Default: null
    sparse-checkout: ''

    # Specifies whether to use cone-mode when doing a sparse checkout.
    # Default: true
    sparse-checkout-cone-mode: ''

    # Number of commits to fetch. 0 indicates all history for all branches and tags.
    # Default: 1
    fetch-depth: ''
```

Instead, it seems to me this behavior is more related to `fetch-depth` setting default value to 1, i.e. fetch only one commit. To get the history until common commit to trunk and trunk-cached, we could set it to zero (fetch all history for all branches and tags, what would be more expansive) or to a considerable number of commits (like 20 or 50) that we believe being less than regular difference between trunk and trunk-cached. WDYT?

**Comment by mumrah** [comment]
Ah, you're right, it's the fetch depth we need. Yes something like 50 or 100 even seems fine.

**Comment by chia7712** [comment]
Pardon me if I misunderstood, but I’m not entirely sure I grasp the use of either the floating tag or trunk-cached. From what I gather, the PR aims to provide a more developer-friendly way to display the cached reference. Developers should consider basing their PRs on the cached reference rather than trunk, as this can leverage the Gradle cache to speed up CI. 

I have two questions if the above description is correct.

1. The downside of basing a PR on a non-trunk branch is that the PR can become out of sync, potentially leading to compile error or failed tests. Additionally, writing code based on a specific cached reference isn’t as straightforward. While I understand the benefit of leveraging the Gradle cache, I’m uncertain if it's worth it and how we could effectively 'request' developers to follow this approach.
2. why not using the [GitHub cache APIs](https://docs.github.com/zh/rest/actions/cache?apiVersion=2022-11-28) to show the latest cached reference? for example:
```
jenkins@25d1ceb7ea06:~/repo$ curl -s "https://api.github.com/repos/apache/kafka/actions/caches?key=gradle-home-v1|Linux-X64|test" | grep "key" | cut -d- -f5 | cut -d'"' -f1 | head -n 1
1bd9dc3ffc78880923a2651aa005a8f564581f9a
```

**Comment by mumrah** [comment]
@chia7712 

> Developers should consider basing their PRs on the cached reference rather than trunk, as this can leverage the Gradle cache to speed up CI.

Yes, this is the main idea here. During busy times when trunk is changing multiple times per day, it will commonly be the case that trunk will be ahead of the cache (since the CI job will be running). So, a developer merging trunk into their branch will be pulling in new changes that aren't in the cache yet. 

I imagine the workflow changing from:

```
git fetch origin
git merge origin/trunk
// resolve conflicts
git push mumrah my-branch
```

to this

```
git fetch origin
git merge origin/trunk-cached
// resolve conflicts
git push mumrah my-branch
```


1. Ideally, the "cached" tag/branch will only be a small number of commits behind trunk. We can advertise this feature in the pull request template. It's not a perfect solution, but for committers (who make lots of changes), we can benefit from fewer cache misses.

2. Do you mean, developers can use this to find the latest cached SHA when merging/rebasing? 
 
edit: I think of this change like an optimization. Not necessary for correctness, but nice to have if you know about it :)

**Comment by mumrah** [comment]
As an example, here's trunk right now

image

Any PRs opened or updated with trunk in the last 3 hours will include 4 commits that won't be included in the cache. Depending on the changes, that is a lot of opportunity for cache misses.

**Comment by chia7712** [comment]
> Do you mean, developers can use this to find the latest cached SHA when merging/rebasing?

Yes, for example: `git merge $SHA`. This approach may not be as straightforward or user-friendly as `trunk-cached`, but it simplifies the workload by leveraging the existing GitHub API. We can document this in the PR template as well.

**Comment by mumrah** [comment]
I actually really like that approach, @chia7712. It's simpler than running additional CI and managing remote refs. 

Just thinking out loud, but perhaps we can use a git alias to find the latest cached SHA and create a local ref. Then we can merge like normal.

```
git update-cache // git alias that uses API to find latest cached SHA, creates a ref
git merge trunk-cached // merge a ref like normal
```

**Comment by chia7712** [comment]
> Just thinking out loud, but perhaps we can use a git alias to find the latest cached SHA and create a local ref. Then we can merge like normal.

+1, I love this idea 😃

**Comment by fonsdant** [comment]
Since it is not possible to push alias to the remote repo, how about to create a little script for this?

**Comment by mumrah** [comment]
@fonsdant yes, exactly. We can create a Git alias that calls a script. The syntax is like:

(in .git/config)
```
[alias]
        test = "!bash test.sh"
```

I think a local branch (i.e., refs/heads/) would make sense. Our script can use `gh cache` to find the latest cached SHA, and then create a local ref using that SHA. It should also print out some info like how many commits behind HEAD the cache is, and how long ago the cache was created.

**Comment by fonsdant** [comment]
@mumrah, @chia7712, I have created an `update-cache.sh` script inside `committer-tools`. It retrieves the latest cached commit (by `created_at`). Also, it makes the Git alias, if it does not already exist. WDYT?

**Comment by mumrah** [comment]
Thanks @fonsdant! I tried it and got an error from `cut`. I'm on an Apple Silicon Mac running zsh

```
$ ./committer-tools/update-cache.sh 
cut: illegal option -- -
usage: cut -b list [-n] [file ...]
       cut -c list [file ...]
       cut -f list [-s] [-w | -d delim] [file ...]

Now, you can use 'git update-cache' as alias to execute this script.
```

**Comment by fonsdant** [comment]
Oh! Okay, I will modify the options, thanks!

**Comment by mumrah** [comment]
Also, I wonder if we should do a `git fetch origin` as part of the command. Otherwise, we might pull in a SHA that doesn't exist locally

**Comment by mumrah** [comment]
Oops, sorry didn't mean to update the branch. `gh update-branch` did the opposite that I expected 😅

**Comment by mumrah** [comment]
The new `cut` options are working 👍 

```
$ git update-cache                 
264131cdaaef3f4696942f26534b3f61f3a2a162
```

**Comment by mumrah** [comment]
Ok, now we need to take the SHA captured by `cut` and use update-ref to create a local branch. 

I'd like to see this output:

```
$ git update-cache
Local branch 'trunk-cached' updated to 264131cdaaef3f4696942f26534b3f61f3a2a162
```

BTW, does the `gh cache` command work for you as a non-committer?

**Comment by fonsdant** [comment]
> Also, I wonder if we should do a git fetch origin as part of the command. Otherwise, we might pull in a SHA that doesn't exist locally

I think it will not work well for all cases. In my local repo, `origin` points to fonsdant/kafka (my fork) and `upstream` to apache/kafka. Maybe, we keep it more agnostic and print a message guiding the user to fetch. WDYT?

> BTW, does the gh cache command work for you as a non-committer?

Yep! :)

```
$ git update-cache
264131cdaaef3f4696942f26534b3f61f3a2a162
```

File: committer-tools/update-cache.sh [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+)
+
+cut -d '-' -f 5 <<< "$key"
+
+if ! git config --get alias.update-cache >/dev/null; then
+  this_path=$(realpath "$0")
+  git config alias.update-cache "!bash $this_path"
+  echo
+  echo "Now, you can use 'git update-cache' as alias to execute this script."
+fi
```
Comment (mumrah) [inline committer-tools/update-cache.sh]: Although I like this feature, it's probably more polite to let committer manually install the script as an alias rather than automatically modifying their git config.

Instead of auto installing it, have this if statement print instructions for adding the alias.

**Comment by mumrah** [comment]
> Maybe, we keep it more agnostic and print a message guiding the user to fetch. WDYT?

Good point on different remote names. Let's still attempt to update the ref in this script. If it can't be done, print a warning and recommend to the user to update their remote.

```
> git update-cache
Cannot update 'trunk-cached' because SHA 264131cdaaef3f4696942f26534b3f61f3a2a162 does not exist locally. Please update your remote and try again.
> echo $? 
1
> git fetch origin-or-whatever
> git update-cache
Local branch 'trunk-cached' updated to 264131cdaaef3f4696942f26534b3f61f3a2a162
```

File: committer-tools/update-cache.sh [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    --key 'gradle-home-v1|Linux-X64|test' \
+    --sort 'created_at' \
+    --limit 1 \
+    --json 'key' \
+    --jq '.[].key'
+)"
+
+sha="$(cut -d '-' -f 5 <<< "$key")"
+
+git fetch origin trunk &> /dev/null
```
Comment (mumrah) [inline committer-tools/update-cache.sh]: as you mentioned, we can't rely on `origin` so lets remove this.

File: committer-tools/update-cache.sh [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+if ! git rev-parse --verify "$sha" &> /dev/null; then
+  printf '\e[33m%s\n%s\e[0m\n' \
+  "Cannot update 'trunk-cached' because SHA $sha" \
+  "does not exist locally. Please update your remote and try again."
+else
+  git switch trunk-cached &> /dev/null || git switch -c trunk-cached &> /dev/null
+  if git merge "$sha" &> /dev/null; then
```
Comment (mumrah) [inline committer-tools/update-cache.sh]: Since this script is managing the `trunk-cached` ref exclusively, lets just use `git update-ref`. 

```
git update-ref -m 'some message' trunk-cached $sha
```

if the SHA doesn't exist, it will fail with a non-zero exit code

File: committer-tools/update-cache.sh [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    printf '%s\n' "Local branch 'trunk-cached' successfully updated to $sha."
```
Comment (mumrah) [inline committer-tools/update-cache.sh]: Can we use a short ref for the SHA and include its relative commit date in this message? 

E.g., 

> Local branch 'trunk-cached' successfully updated to 16cac97873 from 2 hours ago.

File: committer-tools/update-cache.sh (hunk 9 above)
Comment (fonsdant) [reply committer-tools/update-cache.sh]: Oh, thanks! I have mixed it up.

Maybe we could resolve this issue adding the `--all` to `git fetch` as I have pushed. WDYT?

Comment (mumrah) [reply committer-tools/update-cache.sh]: Hmm, I'm not sure about this. I have about 10 kafka remotes in my local Git, so this would be really slow for me 😅 

Try this: 

```
git remote -v | grep 'apache/kafka.git (fetch)' | awk -F" " '{ print $1 }'
```

File: committer-tools/update-cache.sh (hunk 10 above)
Comment (fonsdant) [reply committer-tools/update-cache.sh]: It seems it is not working as expected... I am trying to update to 235cafa805, but my trunk-cached still points to 1f04436de1. See:

```
$ git update-ref -m 'test' trunk-cached 235cafa805
$ echo $?
0
$ git log
...
1f04436de1 (trunk-cached) Merge commit '1854d4b8a11461b53b59fa109b95f2a4f5003997' into trunk-cached
...

Comment (fonsdant) [reply committer-tools/update-cache.sh]: But using `git branch -f` has worked well:

```
$ git branch -f trunk-cached 235cafa805
$ git log
...
235cafa805 (trunk-cached) KAFKA-6197: Update Streams API and Javadoc references in documentation (#17215)
...
```

**Comment by fonsdant** [comment]
The script has failed with `rev-parse` (could not to find commit locally), so I have replaced it with `show`.

File: docs/ops.html [hunk 12]
```diff
@@ -1249,7 +1249,7 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
 
   <h3 class="anchor-heading"><a id="java" class="anchor-link"></a><a href="#java">6.6 Java Version</a></h3>
 
-  Java 8, Java 11, Java 17, and Java 21 are supported.
+  Java 8, Java 11, Java 17, Java 21, and Java 23 are supported.
```
Comment (mumrah) [inline docs/ops.html]: Was this change intentional?

File: committer-tools/update-cache.sh [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    --key 'gradle-home-v1|Linux-X64|test' \
+    --sort 'created_at' \
+    --limit 1 \
+    --json 'key' \
+    --jq '.[].key'
+)"
+
+sha="$(cut -d '-' -f 5 <<< "$key")"
+
+git fetch &> /dev/null
```
Comment (mumrah) [inline committer-tools/update-cache.sh]: The behavior of `git fetch` with no args depends on the local config, so it might not be the same for everyone. 

Since we are expecting developers to have done the appropriate `git fetch` or `git pull` prior to running this command, I think we can omit the `git fetch` completely from this script.

**Review by mumrah — COMMENTED** [review]
Thanks for the updates @fonsdant! I tried it out and it worked as expected. Nice! 

Just two small things then I'm happy to merge this. 

@chia7712 did you have any additional comments?

File: docs/ops.html (hunk 12 above)
Comment (fonsdant) [reply docs/ops.html]: Oh! It should not have been sent actually. Thanks! I will remove it.

File: committer-tools/update-cache.sh (hunk 13 above)
Comment (fonsdant) [reply committer-tools/update-cache.sh]: Make sense, I will proceed this way!

**Review by mumrah — APPROVED** [review]
LGTM!

I'm going to go ahead and merge this after incorporating to two minor nitpicks from the last review. Thanks for sticking with us through this PR @fonsdant 😄. I think we ended up trying three totally different approaches before settling on the local script.

**Comment by chia7712** [comment]
@fonsdant thanks for this great contribution!

**Comment by fonsdant** [comment]
Thank you very much, @mumrah, @chia7712! I have learned a lot along this PR! 😁
